skimage/filters/_multiotsu.pyx

```diff
@@ -0,0 +1,84 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+import numpy as np
+
+cimport numpy as cnp
+cimport cython
+
+
+def _find_threshold_multiotsu(double [:, ::1] var_btwcls, Py_ssize_t classes,
+                              Py_ssize_t bins, double [::1] aux_thresh):
+    """
+    Cython utility function for finding thresholds in multi-Otsu algorithm.
+    This function is only called by filters.threshold_multiotsu.
+
+    Parameters
+    ----------
+
+    var_btwcls : 2-d array
+        array of variance between classes
+    classes ! int
+        number of classes to be segmented
+    bins : int
+        number of bins used in the histogram
+    aux_thresh : array
+        thresholds to be returned
+    """
+    cdef Py_ssize_t idx1, idx2, idx3, idx4
+    cdef double part_sigma = 0
+    cdef double max_sigma = 0
+
+    if classes == 2:
```

Comment:
I guess this is exactly what `combinations` was helpful for :)  @jni Any idea how to do that in Cython?

Reply:
Have we benchmarked using itertools.combinations directly in Cython? And another option is to implement it in Cython, based on the "Python equivalent code", which is in the docstring of combinations:

https://docs.python.org/3/library/itertools.html#itertools.combinations

It could even be modified to modify an input array of Py_ssize_t's inplace. That could be quite fast.